### PR TITLE
Constraint-based TC.

### PR DESCRIPTION
**Pull request by liamoc**
This is a work-in-progress pull request for the new typechecker. It'd be great if zilin could look at it. If anyone else wants to see the code, here it is :)

Currently this breaks the compiler build, because, well, we don't have a typechecker. But the constraint generator is there.

(don't merge this, BTW, at least not until we get this finished)

### CODE REVIEW

**Comment by liamoc** [comment]
I just finished the solver, I think. Now I just need to write the rest of the typechecker infrastructure, hooking it up to the parser, writing pretty printers for the errors (I have some ideas about that that should make them rather nice), etc.

File: cogent/COGENT/Desugar.hs [hunk 1]
```diff
@@ -370,7 +370,10 @@ desugarType t = typeWHNF t >>= \case
   S.RT (S.TUnit)   -> return TUnit
   notInWHNF -> __impossible $ "desugarType" ++ show notInWHNF
 
+-- What is this even doing here?
```
Comment (zilinc) [inline cogent/COGENT/Desugar.hs]: @liamoc same as the whnf function in typechecker. typechecker doesn't convert types to whnf in the output.

**Comment by zilinc** [comment]
@liamoc if you search for `__cogent_debug` in the old impl in `TypeCheck.hs`, you'll see why `tests/fail_ticket-93.cogent` no longer fails.

**Comment by liamoc** [comment]
OK Zilin, I've added the type normalisation phase to the typechecker. The desugarer should no longer need to WHNF types. Also, TypedExprs now have location information, so the desugarer needs to be updated to handle that.

**Comment by liamoc** [comment]
Should I add the _debug functions feature?

**Comment by liamoc** [comment]
Oh, now that I'm a NICTA member should I work directly on the new-tc branch? How do I pull in commit 90f973f to that branch?

**Comment by zilinc** [comment]
@liamoc  Great. I'll fix the desugarer. I don't know whether NICTA membership gives any permissions. The workflow seems to be: everyone develops on the internal repo and only one person can write to the pubic repo, i.e. me for now. Or if you pretend to be external, you can only PR. I don't have admin permission either. I can write, but not changing any settings.

**Comment by zilinc** [comment]
If you can think of any feature better than `_debug`, that'd be better. `_debug` merely works but is far from ideal.

**Comment by zilinc** [comment]
@liamoc it'd be great if you could dev on top of `tc-new` branch so I don't need to merge.

**Comment by zilinc** [comment]
It turns out 90f973f doesn't make life simpler, as now the desugar function needs to preserve normal-form in its inductive definition. But anyway...

**Comment by liamoc** [comment]
But all types are fully normalised before desugaring.. so why do you need to do anything in the desugarer for the types?

**Comment by zilinc** [comment]
Desugaring is inductive: `desugar (take {f1, f2} r)` becomes `desugar (take {f2} r')` so I need to construct `r'`'s type. In the old way, I would just apply a `Take f1` to `r`'s type and `desugar` function is guarded by `typeWHNF`. Now I need think a bit more to update the type of that `f1` field.

**Comment by zilinc** [comment]
Also fixing some bugs that the `normaliseT` function is not a deep one.

**Comment by liamoc** [comment]
Let's close this PR. I'll make new PRs for future changes.

**Pull request closed without merging**